### DISCUSSION TITLE
Recommendation: Don't work against master

### DISCUSSION THREAD

**Discussion by danielb987** [tellurianinteractive/Tellurian.Trains.ModulesRegistryApp#74] [Ideas]
@fjallemark 
I recommend that you don't work directly against the master branch. Early in the project, it sounds like a good idea to work against the master branch, but in the long run it's better to work against other branches and create PRs against master.

If you create branches for your work and create Pull requests to merge those branches into the master branch, other users can see what changes have been done in the project. Often a change consists of several commits and it's the sum of these commits that's of interest. A pull request gives that summary.

If you create Pull requests, these PRs help other to see how to contribute to the project. I have been a software developer for 30 years and worked with several programming languages and I rarely reads a book. I read code and look at how other developers do things. This is one of the main reasons to why I spend so much time review other developers PRs on the JMRI project, since I learn a lot by looking at other developers code.

It makes handling of merge conflicts much easier to handle. If other developers start to create PRs and you decide to give someone else permission to merge PRs, you risk of getting into deep water if you commit directly to the master branch while someone else can merge PRs since you might end up with merge conflicts in the master.
If you instead always commit to a separate branch and creates PRs for your work, GitHub will not allow merging a PR into master if there is a merge conflict. In the JMRI project, merge conflicts happens quite often and GitHub is very good at handling it.

You can work at several things in parallel. Lets say you have a big feature that takes a couple of months to develop. And at the same time, you have bug reports that needs to be resolved. By having one branch for the big feature and another branch for the bug report solution allows you to switch between them.

I have been working on the JMRI project for 3.5 years and have done 126 PRs and it works very well.

**Comment by fjallemark** [comment]
Thanks for the suggestion. There are a number of things to deal with some time in future:
- **Pull requests and branching** as you suggest.
- **Database access** for at least one more database administrator.
- **Azure access** to the resource group for the Module Registry.
- **DevOps** to build and deploy.
- **Database migration strategy**. I use Entity Framework Core but not its migration tooling.

My development philosophy is that not implement something until it becomes necessary for some reason, and not introduce unnecessary complexity, I will therefore wait until a real demand exists for the above items and then solve them. Right now, adding features is more important.
